# 1. Apple's Balance Sheet

### Preparation
We have to search Edgar for Apple Inc. and open its 10-K, and then switch to an HTML rendering of the content.
Then locate its balance sheet, which turns out to be on page 33.

We'll use a snippet from that table as our search pattern

### Data Pulling

In [1]:
import pandas as pd
import requests

In [2]:
URL = 'https://www.sec.gov/Archives/edgar/data/320193/000032019320000096/aapl-20200926.htm'
# I had to experiment with a few different patterns finding this phrase got exactly one result
PATTERN = 'additional paid-in capital,'

In [3]:
html = requests.get(URL).content
result = pd.read_html(html, match=PATTERN)

In [4]:
# did we get a uniquely matching table? yes
len(result)

1

In [5]:
df = result[0]

In [6]:
df.head()

0                          1   \
0                        NaN                        NaN   
1                        NaN                        NaN   
2                    ASSETS:                    ASSETS:   
3            Current assets:            Current assets:   
4  Cash and cash equivalents  Cash and cash equivalents   

                          2                  3                  4   \
0                        NaN                NaN                NaN   
1                        NaN  September 26,2020  September 26,2020   
2                    ASSETS:            ASSETS:            ASSETS:   
3            Current assets:                NaN                NaN   
4  Cash and cash equivalents                  $              38016   

                  5        6        7        8                  9   \
0                NaN      NaN      NaN      NaN                NaN   
1  September 26,2020      NaN      NaN      NaN  September 28,2019   
2            ASSETS:  ASSETS:  ASSETS:  ASSETS:            ASSETS:   
3                NaN      NaN      NaN      NaN                NaN   
4                NaN      NaN      NaN      NaN                  $   

                  10                 11  
0                NaN                NaN  
1  September 28,2019  September 28,2019  
2            ASSETS:            ASSETS:  
3                NaN                NaN  
4              48844                NaN

### Post-processing

In [7]:
# Drop columns with no info or duplicate data
df = df.drop([1, 2, 3, 5, 6, 7, 8, 9, 11], axis=1)
df.head()

0                  4                  10
0                        NaN                NaN                NaN
1                        NaN  September 26,2020  September 28,2019
2                    ASSETS:            ASSETS:            ASSETS:
3            Current assets:                NaN                NaN
4  Cash and cash equivalents              38016              48844

In [8]:
# Rename columns appropriately
df.columns = ['label', 'fy_2020', 'fy_2019']
df

label  \
0                                                 NaN   
1                                                 NaN   
2                                             ASSETS:   
3                                     Current assets:   
4                           Cash and cash equivalents   
5                               Marketable securities   
6                            Accounts receivable, net   
7                                         Inventories   
8                        Vendor non-trade receivables   
9                                Other current assets   
10                               Total current assets   
11                                                NaN   
12                                Non-current assets:   
13                              Marketable securities   
14                 Property, plant and equipment, net   
15                           Other non-current assets   
16                           Total non-current assets   
17                                       Total assets   
18                                                NaN   
19              LIABILITIES AND SHAREHOLDERS’ EQUITY:   
20                               Current liabilities:   
21                                   Accounts payable   
22                          Other current liabilities   
23                                   Deferred revenue   
24                                   Commercial paper   
25                                          Term debt   
26                          Total current liabilities   
27                                                NaN   
28                           Non-current liabilities:   
29                                          Term debt   
30                      Other non-current liabilities   
31                      Total non-current liabilities   
32                                  Total liabilities   
33                                                NaN   
34                      Commitments and contingencies   
35                                                NaN   
36                              Shareholders’ equity:   
37  Common stock and additional paid-in capital, $...   
38                                  Retained earnings   
39      Accumulated other comprehensive income/(loss)   
40                         Total shareholders’ equity   
41         Total liabilities and shareholders’ equity   

                                  fy_2020  \
0                                     NaN   
1                       September 26,2020   
2                                 ASSETS:   
3                                     NaN   
4                                   38016   
5                                   52927   
6                                   16120   
7                                    4061   
8                                   21325   
9                                   11264   
10                                 143713   
11                                    NaN   
12                                    NaN   
13                                 100887   
14                                  36766   
15                                  42522   
16                                 180175   
17                                 323888   
18                                    NaN   
19  LIABILITIES AND SHAREHOLDERS’ EQUITY:   
20                                    NaN   
21                                  42296   
22                                  42684   
23                                   6643   
24                                   4996   
25                                   8773   
26                                 105392   
27                                    NaN   
28                                    NaN   
29                                  98667   
30                                  54490   
31                                 153157   
32                                 258549   
33                                    NaN   
34                   

In [9]:
# The information in the first two rows (date of financial statement) is encoded in the
# column names, so they can be dropped.
df = df.drop([0, 1], axis=0)
df.head()

label  fy_2020  fy_2019
2                    ASSETS:  ASSETS:  ASSETS:
3            Current assets:      NaN      NaN
4  Cash and cash equivalents    38016    48844
5      Marketable securities    52927    51713
6   Accounts receivable, net    16120    22926

In [10]:
# Move the labels into our row index
df = df.set_index('label')
df

fy_2020  \
label                                                                                       
ASSETS:                                                                           ASSETS:   
Current assets:                                                                       NaN   
Cash and cash equivalents                                                           38016   
Marketable securities                                                               52927   
Accounts receivable, net                                                            16120   
Inventories                                                                          4061   
Vendor non-trade receivables                                                        21325   
Other current assets                                                                11264   
Total current assets                                                               143713   
NaN                                                                                   NaN   
Non-current assets:                                                                   NaN   
Marketable securities                                                              100887   
Property, plant and equipment, net                                                  36766   
Other non-current assets                                                            42522   
Total non-current assets                                                           180175   
Total assets                                                                       323888   
NaN                                                                                   NaN   
LIABILITIES AND SHAREHOLDERS’ EQUITY:               LIABILITIES AND SHAREHOLDERS’ EQUITY:   
Current liabilities:                                                                  NaN   
Accounts payable                                                                    42296   
Other current liabilities                                                           42684   
Deferred revenue                                                                     6643   
Commercial paper                                                                     4996   
Term debt                                                                            8773   
Total current liabilities                                                          105392   
NaN                                                                                   NaN   
Non-current liabilities:                                                              NaN   
Term debt                                                                           98667   
Other non-current liabilities                                                       54490   
Total non-current liabilities                                                      153157   
Total liabilities                                                                  258549   
NaN                                                                                   NaN   
Commitments and contingencies                                                         NaN   
NaN                                                                                   NaN   
Shareholders’ equity:                                                                 NaN   
Common stock and additional paid-in capital, $0...                                  50779   
Retained earnings                                                                   14966   
Accumulated other comprehensive income/(loss)                                       (406)   
Total shareholders’ equity                                                          65339   
Total liabilities and shareholders’ equity                                         323888   

                                                                                  fy_2019  
label                                                                              

### Extraction

In [11]:
# Make a list of the fields we want to keep.
keep = ['Total assets', 'Total current assets', 'Total non-current assets',
        'Total liabilities', 'Total current liabilities', 'Total non-current liabilities',
        "Total shareholders’ equity"] # I pasted the special apostrophe directly from copying the balance sheet

In [12]:
# Pull out their values from the DataFrame for Fiscal Year 2020
metrics = df.loc[keep, ['fy_2020']]
metrics

fy_2020
label                                
Total assets                   323888
Total current assets           143713
Total non-current assets       180175
Total liabilities              258549
Total current liabilities      105392
Total non-current liabilities  153157
Total shareholders’ equity      65339

Check the assets, liabilities, and equity add up.

In [13]:
# Remember to convernt numbers to integers! If you forget, the plus sign will just concatenate the strings.
current_assets = int(metrics.loc['Total current assets', 'fy_2020'])
noncurrent_assets = int(metrics.loc['Total non-current assets', 'fy_2020'])
expected = current_assets + noncurrent_assets
actual = int(metrics.loc['Total assets', 'fy_2020'])
expected, actual

(323888, 323888)

In [14]:
current_liab = int(metrics.loc['Total current liabilities', 'fy_2020'])
noncurrent_liab = int(metrics.loc['Total non-current liabilities', 'fy_2020'])
expected = current_liab + noncurrent_liab
actual = int(metrics.loc['Total liabilities', 'fy_2020'])
expected, actual

(258549, 258549)

In [15]:
assets = int(metrics.loc['Total assets', 'fy_2020'])
liab = int(metrics.loc['Total liabilities', 'fy_2020'])
equity = int(metrics.loc['Total shareholders’ equity']) # Note the magic single quote again
assets - liab, equity

(65339, 65339)

Everything looks right! Someone let the SEC know.

# 3. Apple's Cash Flows

### Data Pulling

We'll use the same URL (because this is on the same web page) but a different search pattern.

In [22]:
URL = 'https://www.sec.gov/Archives/edgar/data/320193/000032019320000096/aapl-20200926.htm'
PATTERN = 'cash equivalents and restricted cash,'

In [23]:
html = requests.get(URL).content
result = pd.read_html(html, match=PATTERN)

In [24]:
# did we get a uniquely matching table? yes
len(result)

1

In [25]:
df = result[0]

In [26]:
df

0   \
0                                                 NaN   
1                                                 NaN   
2                                                 NaN   
3   Cash, cash equivalents and restricted cash, be...   
4                               Operating activities:   
5                                          Net income   
6   Adjustments to reconcile net income to cash ge...   
7                       Depreciation and amortization   
8                    Share-based compensation expense   
9                         Deferred income tax benefit   
10                                              Other   
11       Changes in operating assets and liabilities:   
12                           Accounts receivable, net   
13                                        Inventories   
14                       Vendor non-trade receivables   
15               Other current and non-current assets   
16                                   Accounts payable   
17                                   Deferred revenue   
18          Other current and non-current liabilities   
19             Cash generated by operating activities   
20                              Investing activities:   
21                 Purchases of marketable securities   
22  Proceeds from maturities of marketable securities   
23       Proceeds from sales of marketable securities   
24  Payments for acquisition of property, plant an...   
25  Payments made in connection with business acqu...   
26             Purchases of non-marketable securities   
27            Proceeds from non-marketable securities   
28                                              Other   
29   Cash generated by/(used in) investing activities   
30                              Financing activities:   
31             Proceeds from issuance of common stock   
32  Payments for taxes related to net share settle...   
33    Payments for dividends and dividend equivalents   
34                        Repurchases of common stock   
35           Proceeds from issuance of term debt, net   
36                            Repayments of term debt   
37                Repayments of commercial paper, net   
38                                              Other   
39                  Cash used in financing activities   
40  Increase/(Decrease) in cash, cash equivalents ...   
41  Cash, cash equivalents and restricted cash, en...   
42                 Supplemental cash flow disclosure:   
43                    Cash paid for income taxes, net   
44                             Cash paid for interest   

                                                   1   \
0                                                 NaN   
1                                                 NaN   
2                                                 NaN   
3   Cash, cash equivalents and restricted cash, be...   
4                               Operating activities:   
5                                          Net income   
6   Adjustments to reconcile net income to cash ge...   
7                       Depreciation and amortization   
8                    Share-based compensation expense   
9                         Deferred income tax benefit   
10                                              Other   
11       Changes in operating assets and liabilities:   
12                           Accounts receivable, net   
13                                        Inventories   
14                       Vendor non-trade receivables   
15               Other current and non-current assets   
16                                   Accounts payable   
17                                   Deferred revenue   
18          Other current and non-current liabilities   
19             Cash generated by operating activities   
20                              Investing activities:   
21                 Purchases of marketable securities   
22  Proceeds from maturities of marketable securities   
23       Proceeds from sales of marketable securities   
24 

### Post-processing

Lots of useless columns in this one -- everything except 0, 4, 10, and 16.

Instead of dropping all the other columns, we can use selection syntax to just keep those four.

In [27]:
df = df[[0, 4, 10, 16]]
df.head(10)

0                  4   \
0                                                NaN                NaN   
1                                                NaN        Years ended   
2                                                NaN  September 26,2020   
3  Cash, cash equivalents and restricted cash, be...              50224   
4                              Operating activities:                NaN   
5                                         Net income              57411   
6  Adjustments to reconcile net income to cash ge...                NaN   
7                      Depreciation and amortization              11056   
8                   Share-based compensation expense               6829   
9                        Deferred income tax benefit              (215)   

                  10                 16  
0                NaN                NaN  
1        Years ended        Years ended  
2  September 28,2019  September 29,2018  
3              25913              20289  
4                NaN                NaN  
5              55256              59531  
6                NaN                NaN  
7              12547              10903  
8               6068               5340  
9              (340)           (32,590)

Make the column names into the fiscal years.

In [28]:
df.columns = ['label', 'fy_2020', 'fy_2019', 'fy_2018']
df.head(10)

label            fy_2020  \
0                                                NaN                NaN   
1                                                NaN        Years ended   
2                                                NaN  September 26,2020   
3  Cash, cash equivalents and restricted cash, be...              50224   
4                              Operating activities:                NaN   
5                                         Net income              57411   
6  Adjustments to reconcile net income to cash ge...                NaN   
7                      Depreciation and amortization              11056   
8                   Share-based compensation expense               6829   
9                        Deferred income tax benefit              (215)   

             fy_2019            fy_2018  
0                NaN                NaN  
1        Years ended        Years ended  
2  September 28,2019  September 29,2018  
3              25913              20289  
4                NaN                NaN  
5              55256              59531  
6                NaN                NaN  
7              12547              10903  
8               6068               5340  
9              (340)           (32,590)

That makes the first 3 rows obselete...

In [29]:
df = df.drop([0, 1, 2], axis=0)
df.head(10)

label fy_2020 fy_2019  \
3   Cash, cash equivalents and restricted cash, be...   50224   25913   
4                               Operating activities:     NaN     NaN   
5                                          Net income   57411   55256   
6   Adjustments to reconcile net income to cash ge...     NaN     NaN   
7                       Depreciation and amortization   11056   12547   
8                    Share-based compensation expense    6829    6068   
9                         Deferred income tax benefit   (215)   (340)   
10                                              Other    (97)   (652)   
11       Changes in operating assets and liabilities:     NaN     NaN   
12                           Accounts receivable, net    6917     245   

     fy_2018  
3      20289  
4        NaN  
5      59531  
6        NaN  
7      10903  
8       5340  
9   (32,590)  
10     (444)  
11       NaN  
12   (5,322)

Move the labels to be row indices.

In [30]:
df = df.set_index('label')
df.head(10)

fy_2020 fy_2019   fy_2018
label                                                                       
Cash, cash equivalents and restricted cash, beg...   50224   25913     20289
Operating activities:                                  NaN     NaN       NaN
Net income                                           57411   55256     59531
Adjustments to reconcile net income to cash gen...     NaN     NaN       NaN
Depreciation and amortization                        11056   12547     10903
Share-based compensation expense                      6829    6068      5340
Deferred income tax benefit                          (215)   (340)  (32,590)
Other                                                 (97)   (652)     (444)
Changes in operating assets and liabilities:           NaN     NaN       NaN
Accounts receivable, net                              6917     245   (5,322)

### Extraction
We have what we need now -- let's just look at the fields we're interested in.

In [31]:
# It can be helpful to look at the labels we'll have to pick from. Remember, they're
# stored in the (row) index now.
df.index

Index(['Cash, cash equivalents and restricted cash, beginning balances',
       'Operating activities:', 'Net income',
       'Adjustments to reconcile net income to cash generated by operating activities:',
       'Depreciation and amortization', 'Share-based compensation expense',
       'Deferred income tax benefit', 'Other',
       'Changes in operating assets and liabilities:',
       'Accounts receivable, net', 'Inventories',
       'Vendor non-trade receivables', 'Other current and non-current assets',
       'Accounts payable', 'Deferred revenue',
       'Other current and non-current liabilities',
       'Cash generated by operating activities', 'Investing activities:',
       'Purchases of marketable securities',
       'Proceeds from maturities of marketable securities',
       'Proceeds from sales of marketable securities',
       'Payments for acquisition of property, plant and equipment',
       'Payments made in connection with business acquisitions, net',
       'Purcha

We can copy and paste label names from above so we know we get it exactly right.

#### Operating Cash

In [32]:
# Change in cash from operating activities
operating_label = 'Cash generated by operating activities'
delta_operating = int(df.loc[operating_label, 'fy_2020']) - int(df.loc[operating_label, 'fy_2019'])
delta_operating

11283

#### Investing Cash

In [33]:
# Change in cash from investing activities
investing_label = 'Cash generated by/(used in) investing activities'
delta_investing = int(df.loc[investing_label, 'fy_2020']) - int(df.loc[investing_label, 'fy_2019'])
delta_investing

ValueError: invalid literal for int() with base 10: '(4,289)'

Uh oh! We ran into a number in parenthesis, which can't automatically be converted into an integer.
Parentheses indicate negative numbers in financial statements, so we'll have to work more carefully with this one.

In [34]:
investing_2020 = df.loc[investing_label, 'fy_2020']
investing_2019 = df.loc[investing_label, 'fy_2019']
investing_2020, investing_2019

('(4,289)', '45896')

So the first is negative. 
We could change this by hand (just compare `-4289` to `45896`), but a more elegant way would be to do it programmatically.

In [35]:
def normalize_number(number):
    '''
    Takes a number as a string and returns it as an integer. If the number has parens,
    it will be returned as a negative.
    '''
    # Turn parenthetical numbers into negatives
    if '(' in number:
        # Number is a string, so we can strip characters and add a negative sign at the front.
        number = number.strip('()')
        number = '-' + number
    return int(number)

Now we can call our function on **any** number we find in a financial statement, to safely convert it to a positive or negative integer.

In [36]:
normalize_number(investing_2020), normalize_number(investing_2019)

ValueError: invalid literal for int() with base 10: '-4,289'

Oh! And now we have to deal with commas too, which Python doesn't like. Let's modify our function slightly.

In [37]:
def normalize_number(number):
    '''
    Takes a number as a string and returns it as an integer. If the number has parens,
    it will be returned as a negative.
    '''
    # Turn parenthetical numbers into negatives
    if '(' in number:
        # Number is a string, so we can strip characters and add a negative sign at the front.
        number = number.strip('()')
        number = '-' + number
    # Remove commas
    number = number.replace(',', '')
    return int(number)

In [38]:
normalize_number(investing_2020), normalize_number(investing_2019)

(-4289, 45896)

In [39]:
delta_investing = normalize_number(investing_2020) - normalize_number(investing_2019)
delta_investing

-50185

#### Financing Cash

In [40]:
# Change in cash from financing activities
financing_label = 'Cash used in financing activities'
delta_financing = int(df.loc[financing_label, 'fy_2020']) - int(df.loc[financing_label, 'fy_2019'])
delta_financing

ValueError: invalid literal for int() with base 10: '(86,820)'

Looks like we need our function again! The great thing about functions is that you can reuse them.

In [41]:
delta_financing = normalize_number(df.loc[financing_label, 'fy_2020']) - normalize_number(df.loc[financing_label, 'fy_2019'])
delta_financing

4156

#### Comparing Them

In [42]:
print('Change in operating cash:', delta_operating)
print('Change in investing cash:', delta_investing)
print('Change in financing cash:', delta_financing)

Change in operating cash: 11283
Change in investing cash: -50185
Change in financing cash: 4156


So the greatest change from 2019 to 2020 was in Apple's investment, which decreased by $50 billion. Interesting.
We can even look at the amounts.

In [43]:
investing_2020, investing_2019

('(4,289)', '45896')

So they went from *making* $46b on investments to *losing* 5 billion. Quite a swing.